esp-hal-common/src/uart.rs

```diff
@@ -1390,116 +1393,163 @@ mod asynch {
+                        match event {
+                            RxEvent::RxFifoFull => w.rxfifo_full_int_ena().set_bit(),
+                            RxEvent::RxCmdCharDetected => w.at_cmd_char_det_int_ena().set_bit(),
+                            RxEvent::RxFifoOvf => w.rxfifo_ovf_int_ena().set_bit(),
+                            RxEvent::RxFifoTout => w.rxfifo_tout_int_ena().set_bit(),
+                        };
+                    }
+                    w
+                });
+                self.registered = true;
+            }
+            if self.event_bit_is_clear() {
+                Poll::Ready(())
+            } else {
+                Poll::Pending
+            }
+        }
     }
 
-    impl<'d, T: Instance> UartFuture<'d, T> {
-        pub fn new(event: Event) -> Self {
-            match event {
-                Event::TxDone => T::register_block()
-                    .int_ena
-                    .modify(|_, w| w.tx_done_int_ena().set_bit()),
-                Event::TxFiFoEmpty => T::register_block()
-                    .int_ena
-                    .modify(|_, w| w.txfifo_empty_int_ena().set_bit()),
-                Event::RxFifoFull => T::register_block()
-                    .int_ena
-                    .modify(|_, w| w.rxfifo_full_int_ena().set_bit()),
-                Event::RxCmdCharDetected => T::register_block()
-                    .int_ena
-                    .modify(|_, w| w.at_cmd_char_det_int_ena().set_bit()),
-                Event::RxFifoOvf => T::register_block()
-                    .int_ena
-                    .modify(|_, w| w.rxfifo_ovf_int_ena().set_bit()),
-                Event::RxFifoTout => T::register_block()
-                    .int_ena
-                    .modify(|_, w| w.rxfifo_tout_int_ena().set_bit()),
+    impl<'d, T: Instance> Drop for UartRxFuture<'d, T> {
+        fn drop(&mut self) {
+            // Although the isr disables the interrupt that occured directly, we need to
+            // disable the other interrupts (= the ones that did not occur), as
+            // soon as this future goes out of scope.
+            let int_ena = &T::register_block().int_ena;
+            for event in self.events {
+                match event {
+                    RxEvent::RxFifoFull => {
+                        int_ena.modify(|_, w| w.rxfifo_full_int_ena().clear_bit())
+                    }
+                    RxEvent::RxCmdCharDetected => {
+                        int_ena.modify(|_, w| w.at_cmd_char_det_int_ena().clear_bit())
+                    }
+                    RxEvent::RxFifoOvf => int_ena.modify(|_, w| w.rxfifo_ovf_int_ena().clear_bit()),
+                    RxEvent::RxFifoTout => {
+                        int_ena.modify(|_, w| w.rxfifo_tout_int_ena().clear_bit())
+                    }
+                }
             }
+        }
+    }
 
+    impl<'d, T: Instance> UartTxFuture<'d, T> {
+        pub fn new(events: &'d Vec<TxEvent, 2>) -> Self {
```

Comment:
@MabezDev, seems like esp_idf_hal but not sdp_hal_common, is depending on enumset, which would more exactly what is needed here than a heapless vec.
 Is there something similar used in esp_hal instead of enumset, or would it be ok to add enumset as a dependency to be able to work with a set of enums instead of a vector of enums here?

Reply:
`enumset` would be fine to add as a dep I think.